File: w3/dz_cpp/test/template/point.cpp
#include <iostream>

template <typename T>
class Point {
    private:
        T x, y;

    public:
        Point() {
            this->x = 0;
            this->y = 0;
        }

        Point(T x, T y) {
            this->x = x;
            this->y = y;
        }

        T getX() const {
            return this->x;
        }

        T getY() const {
            return this->y;
        }

        void print() {
            std::cout << "(" << this->x << ", " << this->y << ")" << std::endl;
        }
};

int main() {
    Point<float> *p1 = new Point<float>(2.2, 3.5);
    Point<int> *p2 = new Point<int>(4, 2);
    Point<double> *p3 = new Point<double>(1.1, 9.1);

    p1->print();
    p2->print();
    p3->print();

    return 0;
}

File: w3/dz_java/src/Car/Point.java
import static java.lang.Math.hypot;

public class Point {
    private double x;
    private double y;

    public Point() {
        this.x = 0;
        this.y = 0;
    }

    public Point (double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double distance(Point other) {
        return hypot(this.x - other.x, this.y - other.y);
    }

    boolean equal(Point other) {
        return this.x == other.x && this.y == other.y;
    }

    boolean notEqual(Point other) {
        return !this.equal(other);
    }

    @Override
    public String toString() {
        return String.format("(" + this.getX() + ", " + this.getY() + ")");
    }
}
